setup_support.py

```diff
@@ -100,7 +108,7 @@ def run(self):
                 cmdname = "{0}{1}".format(
                     'datalad ' if cmdname != 'datalad' else '',
                     cmdname)
-                format = cls(cmdname, ext_sections=sections, version=get_version())
+                format = cls(cmdname, ext_sections=sections, version=get_version(appname))
```

Comment:
Wouldn't this result in the manpages for extensions having the wrong version (datalad's rather than the extension's)?

Reply:
good catch!
- apparently we do not yet ship manpages for extensions' commands (I see none in datalad-container debian package at least, and `datalad` package itself doesn't ship the ones for extensions)
- I don't even know if they build fine in extensions ;)
- even if we did we stick to hardcoded "datalad" in those copies of setup_support.py
- indeed should be fixed properly here and better to provide the version of extension here than of datalad.
- ... I will look into checking/adding support for it later (not today) unless someone beats me to it

Reply:
> * I don't even know if they build fine in extensions ;)

Based on the [-container rtd page][0] and on running `python setup.py build_manpage` locally, I think they build fine.

[0]: https://datalad-container.readthedocs.io/en/latest/generated/man/datalad-containers-add.html

> * even if we did we stick to hardcoded "datalad" in those copies of setup_support.py

We stick to a hard-coded `appname` of "datalad" in `BuildManPage`, but `get_version` does appear to be adjusted for each extension that has setup_support.py:

https://github.com/datalad/datalad-container/blob/8cf9f9b250d58695a38b3335f59fe667566ff375/setup_support.py#L28-L34

So the current state is that an extension uses the extension's version in `BuildManPage`.  Running `python setup.py build_manpage` for -container generated a manpage having the expected footer:

```
datalad-containers-add 0.5.0                  2019-08-14                     datalad-containers-add(1)
```

Looking more closely at `get_version`, I actually think passing "datalad" to `get_version` would fail for extensions because `get_version` would expect "./datalad/version.py" to exist:

https://github.com/datalad/datalad/blob/652cd4d9803a6af2b0d0ce9703ddb71bd3461247/setup_support.py#L39-L40

Reply:
Thanks for the analysis!

Reply:
ok, pushed somewhat ugly (not worse than IMHO some other `datalad.extensions` handling points) but seems to be working workaround for all of this.  Now datalad-container does get their manpages generated properly
```shell
(git)hopa:~datalad/datalad-container[debian]git
$> rm -rf build/man && python setup.py build_manpage && for f in ./build/man/*.*; do head -n 1 $f; done
running build_manpage
.TH datalad containers-add 1 2019\-08\-14 "datalad containers-add 0.5.0"
.TH datalad containers-list 1 2019\-08\-14 "datalad containers-list 0.5.0"
.TH datalad containers-remove 1 2019\-08\-14 "datalad containers-remove 0.5.0"
.TH datalad containers-run 1 2019\-08\-14 "datalad containers-run 0.5.0"
```
with its setup.py being
```python
#!/usr/bin/env python

from setup_support import (
    BuildManPage,
    BuildRSTExamplesFromScripts,
    datalad_setup,
)


cmdclass = {
    'build_manpage': BuildManPage,
    'build_examples': BuildRSTExamplesFromScripts,
}


datalad_setup(
    # basic project properties can be set arbitrarily
    name="datalad_container",
    description="DataLad extension package for working with containerized environments",
    # datalad command suite specs from here
    install_requires=[
        'datalad>=0.11.5',
        'requests>=1.2',  # to talk to Singularity-hub
    ],
    extras_require={
        'devel-docs': [
            # used for converting README.md -> .rst for long_description
            'pypandoc',
            # Documentation
            'sphinx',
            'sphinx-rtd-theme',
        ]},
    cmdclass=cmdclass,
    entry_points = {
        # 'datalad.extensions' is THE entrypoint inspected by the datalad API builders
        'datalad.extensions': [
            # the label in front of '=' is the command suite label
            # the entrypoint can point to any symbol of any name, as long it is
            # valid datalad interface specification (see demo in this extension)
            'container=datalad_container:command_suite',
        ],
        'datalad.tests': [
            'container=datalad_container',
        ],
    },
)
```
which seems to be quite minimalistic

Reply:
done

Reply:
I can confirm that your latest changes add the correct version to the manpage.  Thanks!

I did notice that the manpage looks a bit off.  A diff with your changes as the new version:

```diff
--- out.0	2019-08-15 10:08:07.731791152 -0400
+++ out.1	2019-08-15 10:08:19.400217070 -0400
@@ -1,7 +1,7 @@
-datalad-containers-list(1)              General Commands Manual             datalad-containers-list(1)
+datalad(containers-list)             datalad containers-list 0.5.0            datalad(containers-list)
 
 SYNOPSIS
-       datalad-containers-list [-h] [-d DATASET] [-r]
+       datalad containers-list [-h] [-d DATASET] [-r]
 
 DESCRIPTION
        List containers known to a dataset
@@ -22,4 +22,4 @@
 AUTHORS
         datalad is developed by The DataLad Team and Contributors <team@datalad.org>.
 
-datalad-containers-list 0.5.0                 2019-08-15                    datalad-containers-list(1)
+2019-08-15                                         1                          datalad(containers-list)
```

The synopsis change looks like an improvement because a hyphenated `datalad-containers-list` isn't valid, but the other lines suggest that `container-list` is taken as the section.

Perhaps we should update `_mk_title` to do something like this?

```diff
diff --git a/formatters.py b/formatters.py
index 3a997db..1b93e52 100644
--- a/formatters.py
+++ b/formatters.py
@@ -69,8 +69,9 @@ def _mk_synopsis(self, parser):
         return usage
 
     def _mk_title(self, prog):
-        name_version = "\"{0} {1}\"".format(prog, self._version)
-        return '.TH {0} {1} {2} {3}\n'.format(prog, self._section,
+        prog_hyph = "-".join(prog.split())
+        name_version = "\"{0} {1}\"".format(prog_hyph, self._version)
+        return '.TH {0} {1} {2} {3}\n'.format(prog_hyph, self._section,
                                               self._today, name_version)
 
     def _make_name(self, parser):
```

Diffing with the output before this PR, that gives

```diff
--- out.0	2019-08-15 10:08:07.731791152 -0400
+++ out.2	2019-08-15 10:08:26.224466367 -0400
@@ -1,7 +1,7 @@
 datalad-containers-list(1)              General Commands Manual             datalad-containers-list(1)
 
 SYNOPSIS
-       datalad-containers-list [-h] [-d DATASET] [-r]
+       datalad containers-list [-h] [-d DATASET] [-r]
 
 DESCRIPTION
        List containers known to a dataset
```

Reply:
`<rant>`
oh crap - there is also formatters.py which is copied and changed around - I have not even looked into that one yet, and it does have difference to version (probably just outdated) in extensions ... 
It does feel that we are just reaching the point of making `datalad-support` (or what other name?) package now since situation is really far from optimal :-/  BUT we actually cannot since setuptools doesn't have a way to declare "build dependencies" . IIRC it is some kind of a new builder (can't find ATM) has it, and there is https://www.python.org/dev/peps/pep-0518/ and https://github.com/pypa/pipenv/issues/3651 .  But it feels that even if we make such a package we end up just copying it around.  At least may be it would get proper versioning information to check were it is outdated etc.
`</rant>`

Reply:
> Diffing with the output before this PR, that gives
> 
> ```diff
> --- out.0	2019-08-15 10:08:07.731791152 -0400
> +++ out.2	2019-08-15 10:08:26.224466367 -0400
> @@ -1,7 +1,7 @@
>  datalad-containers-list(1)              General Commands Manual            
> ```

I am confused a bit -- so it reverts back to suboptimal?

If I symlink formatters.py from datalad core as well (no changes to it there), then things seems to be rendered ok for me:

```shell
(git)hopa:~datalad/datalad-container[debian]git
$> rm -rf build/man && python setup.py build_manpage && for f in ./build/man/*.*; do man $f | head -n 8; done 
running build_manpage
datalad containers-add(1)                                            datalad                                            datalad containers-add(1)

NAME
       datalad containers-add - add a container to a dataset

SYNOPSIS
       datalad containers-add [-h] [-u URL] [-d DATASET] [--call-fmt FORMAT] [-i IMAGE] [--update] NAME

datalad containers-list(1)                                           datalad                                           datalad containers-list(1)

NAME
       datalad containers-list - list containers known to a dataset

SYNOPSIS
       datalad containers-list [-h] [-d DATASET] [-r]

datalad containers-remove(1)                                         datalad                                         datalad containers-remove(1)

NAME
       datalad containers-remove - remove a known container from a dataset

SYNOPSIS
       datalad containers-remove [-h] [-d DATASET] [-i] NAME

datalad containers-run(1)                                            datalad                                            datalad containers-run(1)

NAME
       datalad containers-run - drop-in replacement of 'run' to perform containerized command execution

SYNOPSIS
       datalad containers-run [-h] [-n NAME] [-d DATASET] [-i PATH] [-o PATH] [-m MESSAGE] [--expand WHICH] [--explicit] [--sidecar yes|no] ...
```

Reply:
> I am confused a bit -- so it reverts back to suboptimal?

The main point I was trying to make is that, with your setup_support.py change and my suggested tweak to `_mk_title`, running `python setup.py build_manpage` from the -container repo produces the same rendered manpage output, aside from the more correct synopsis line.  Without the `_mk_title` tweak, there is the weirdness with "containers-list" being formatted as the section.

> If I symlink formatters.py from datalad core as well (no changes to it there), then things seems to be rendered ok for me:

Good catch!  formatters.py from datalad and -container have indeed diverged (woohoo for copying files :):

```
# -container repo
% git describe
0.5.0
% git diff --no-index --stat formatters.py ../datalad/formatters.py
 formatters.py => ../datalad/formatters.py | 18 +++++++++++-------
 1 file changed, 11 insertions(+), 7 deletions(-)
```

I haven't inspected the differences, but using datalad core's version of formatters.py from the -container repo, I don't see the "containter-list as section" issue.  So, as your comment suggests, all seems good there!

(With the datalad master version of formatter.py, I'm not seeing the version in the footer.  I'm assuming that is an unintended change,   but the behavior is shared by both datalad master and this PR, so I don't think we should worry about it in this PR.)

Reply:
> (With the datalad master version of formatter.py, I'm not seeing the version in the footer. I'm assuming that is an unintended change, but the behavior is shared by both datalad master and this PR, so I don't think we should worry about it in this PR.)

NICE CATCH!  reproduced.  I would just fix here -- will look later at what is the diff -- I bet we broke it with those formatters changes in the core which aren't in the extensions' copy